zcash_primitives/src/block.rs

```diff
@@ -0,0 +1,209 @@
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use hex;
+use std::fmt;
+use std::io::{self, Read, Write};
+use std::ops::Deref;
+
+use serialize::Vector;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct BlockHash(pub [u8; 32]);
+
+impl fmt::Display for BlockHash {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        let mut data = self.0.clone();
+        data.reverse();
+        formatter.write_str(&hex::encode(data))
+    }
+}
+
+/// A Zcash block header.
+pub struct BlockHeader(BlockHeaderData);
+
+impl Deref for BlockHeader {
+    type Target = BlockHeaderData;
+
+    fn deref(&self) -> &BlockHeaderData {
+        &self.0
+    }
+}
+
+pub struct BlockHeaderData {
+    pub version: i32,
+    pub prev_block: BlockHash,
+    pub merkle_root: [u8; 32],
+    pub final_sapling_root: [u8; 32],
+    pub time: u32,
+    pub bits: u32,
+    pub nonce: [u8; 32],
+    pub solution: Vec<u8>,
+}
+
+impl BlockHeaderData {
+    pub fn freeze(self) -> BlockHeader {
+        BlockHeader(self)
+    }
+}
+
+impl BlockHeader {
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let version = reader.read_i32::<LittleEndian>()?;
+
+        let mut prev_block = BlockHash([0; 32]);
+        reader.read_exact(&mut prev_block.0)?;
+
+        let mut merkle_root = [0; 32];
+        reader.read_exact(&mut merkle_root)?;
+
+        let mut final_sapling_root = [0; 32];
+        reader.read_exact(&mut final_sapling_root)?;
```

Comment:
I don't think there's anything wrong with this code, but I get a little anxious about the `let mut bytes = zeros; read bytes;` pattern, since (considering the reading as the "real" initialization) the variable initialization is unlinked from its declaration, and the resulting variable is left mutable even if it doesn't need to be.

Since reading 32 bytes is a pretty common idiom, maybe it would make sense to define some extension behaviour like so:
```rust
pub trait ReadExt {
    fn read_32_bytes(&mut self) -> io::Result<[u8; 32]>;
}

impl<R: Read> ReadExt for R {
    #[inline]
    fn read_32_bytes(&mut self) -> io::Result<[u8; 32]> {
        let mut bytes = [0; 32];
        self.read_exact(&mut bytes)?;
        Ok(bytes)
    }
}
```

so that this byte-handling code can become, e.g.,
```rust
let prev_block = BlockHash(reader.read_32_bytes()?);
let merkle_root = reader.read_32_bytes()?;
let final_sapling_root = reader.read_32_bytes()?;
```

Reply:
This is a commonly-used idiom throughout the existing codebase, so we should address this comprehensively. Opened https://github.com/zcash/librustzcash/issues/72